### PR TITLE
feat: add predicate example and docs

### PR DESCRIPTION
**Pull request by hal3e**
closes: https://github.com/FuelLabs/fuels-rs/issues/535

I have added a simple predicate example which requires the user to send two out of three signatures that match some hard coded public keys.

### CODE REVIEW

File: docs/src/getting-started/predicates.md [hunk 1]
```diff
@@ -0,0 +1,15 @@
+# Predicates
+
+Predicates, in Sway, are programs that return a Bolean value and they do not have any side effects (they are pure).
```
Comment (Br1ght0ne) [inline docs/src/getting-started/predicates.md:3]:
```suggestion
Predicates, in Sway, are programs that return a Boolean value and they do not have any side effects (they are pure).
```

File: docs/src/getting-started/predicates.md [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+## Instantiating predicates
+
+Similarity to contracts, once you've written a predicate in Sway and compiled it with `forc build` (read [here](https://fuellabs.github.io/sway/master/introduction/overview.html) for more on how to work with Sway), you'll get the predicate binary. Using the binary, you can instantiate an `predicate` as shown in the code snippet below:
```
Comment (Br1ght0ne) [inline docs/src/getting-started/predicates.md:7]:
```suggestion
Similarly to contracts, once you've written a predicate in Sway and compiled it with `forc build` (read [here](https://fuellabs.github.io/sway/master/introduction/overview.html) for more on how to work with Sway), you'll get the predicate binary. Using the binary, you can instantiate an `predicate` as shown in the code snippet below:
```

**Review by Br1ght0ne — CHANGES_REQUESTED** [review]
Other than a few typos, looks wonderful.

File: docs/src/getting-started/predicates.md [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+
+```rust,ignore
+{{#include ../../../examples/predicates/src/lib.rs:predicate_load_from}}
+```
+The created `predicate` instance has two fields. The predicate `byte code` and the predicate `address`. This address is generated from the byte code and is the same as the `P2SH` address used in Bitcoin. Users can seamlessly send assets to the predicate address as they do for any other address on the chain. To spend the predicate funds, the user has to provide the original `byte code` of the predicate together with the `predicate data`. The `predicate data` will be used when executing the `byte code` and if the predicate is validated successfully the funds will be spent.
+
+In the next section, we will show in detail how to interact with a predicate and explore an example where specific signatures are needed to spend the predicate funds.
```
Comment (MujkicA) [inline docs/src/getting-started/predicates.md:14]: Personal preference, take it or leave it
```suggestion
In the next section, we show how to interact with a predicate and explore an example where specific signatures are needed to spend the predicate funds.
```

File: docs/src/getting-started/predicates.md (hunk 1 above)
Comment (MujkicA) [inline docs/src/getting-started/predicates.md:3]:
```suggestion
Predicates, in Sway, are programs that return a Boolean value, and they do not have any side effects (they are pure).
```

File: docs/src/getting-started/predicates.md (hunk 2 above)
Comment (MujkicA) [inline docs/src/getting-started/predicates.md:7]:
```suggestion
Similar to contracts, once you've written a predicate in Sway and compiled it with `forc build` (read [here](https://fuellabs.github.io/sway/master/introduction/overview.html) for more on how to work with Sway), you'll get the predicate binary. Using the binary, you can instantiate a `predicate` as shown in the code snippet below:
```

File: docs/src/getting-started/predicates.md [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+Predicates, in Sway, are programs that return a Bolean value and they do not have any side effects (they are pure).
+
+## Instantiating predicates
+
+Similarity to contracts, once you've written a predicate in Sway and compiled it with `forc build` (read [here](https://fuellabs.github.io/sway/master/introduction/overview.html) for more on how to work with Sway), you'll get the predicate binary. Using the binary, you can instantiate an `predicate` as shown in the code snippet below:
+
+```rust,ignore
+{{#include ../../../examples/predicates/src/lib.rs:predicate_load_from}}
+```
+The created `predicate` instance has two fields. The predicate `byte code` and the predicate `address`. This address is generated from the byte code and is the same as the `P2SH` address used in Bitcoin. Users can seamlessly send assets to the predicate address as they do for any other address on the chain. To spend the predicate funds, the user has to provide the original `byte code` of the predicate together with the `predicate data`. The `predicate data` will be used when executing the `byte code` and if the predicate is validated successfully the funds will be spent.
```
Comment (MujkicA) [inline docs/src/getting-started/predicates.md:12]:
```suggestion
The created `predicate` instance has two fields. The predicate `byte code` and the predicate `address`. This address is generated from the byte code and is the same as the `P2SH` address used in Bitcoin. Users can seamlessly send assets to the predicate address as they do for any other address on the chain. To spend the predicate funds, the user has to provide the original `byte code` of the predicate together with the `predicate data`. The `predicate data` will be used when executing the `byte code`, and if the predicate is validated successfully, the funds will be accessible.
```

File: docs/src/predicates/send-spend-predicate.md [hunk 5]
```diff
@@ -0,0 +1,45 @@
+# Send and spend funds from predicates
+
+Let's consider the following predicate example:
+
+```rust,ignore
+{{#include ../../../packages/fuels/tests/test_projects/predicate_signatures/src/main.sw}}
+```
+
+This predicate accepts three signatures and matches them to three predefined public keys. The `ec_recover_address` function is used to recover the public key from the signatures. If two of three extracted public keys match the predefined public keys the funds can be spent. Note that the signature order has to match the order of the predefined public keys.
```
Comment (MujkicA) [inline docs/src/predicates/send-spend-predicate.md:9]:
```suggestion
This predicate accepts three signatures and matches them to three predefined public keys. The `ec_recover_address` function is used to recover the public key from the signatures. If two of three extracted public keys match the predefined public keys, the funds can be spent. Note that the signature order has to match the order of the predefined public keys.
```

File: docs/src/predicates/send-spend-predicate.md [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+{{#include ../../../examples/predicates/src/lib.rs:predicate_coins}}
+```
+
+Now we can load the predicate binary, and prepare variables.
+
+```rust,ignore
+{{#include ../../../examples/predicates/src/lib.rs:predicate_load}}
+```
+
+After the predicate address is generated we can send funds to it. Note that we are using the same `tranfer` function as we used when sending funds to other wallets. We also make sure that the funds are indeed transferred.
```
Comment (MujkicA) [inline docs/src/predicates/send-spend-predicate.md:29]:
```suggestion
After the predicate address is generated we can send funds to it. Note that we are using the same `transfer` function as we used when sending funds to other wallets. We also make sure that the funds are indeed transferred.
```

File: docs/src/predicates/send-spend-predicate.md [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+
+```rust,ignore
+{{#include ../../../examples/predicates/src/lib.rs:predicate_send}}
+```
+
+To spend the funds that are now locked in the predicate we have to provide two out of three signatures whose public keys match the ones we defined in the predicate. The signatures are generate in the snippet below:
```
Comment (MujkicA) [inline docs/src/predicates/send-spend-predicate.md:35]:
```suggestion
To spend the funds that are now locked in the predicate we have to provide two out of three signatures whose public keys match the ones we defined in the predicate. The signatures are generated in the snippet below:
```

File: docs/src/predicates/send-spend-predicate.md [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+
+```rust,ignore
+{{#include ../../../examples/predicates/src/lib.rs:predicate_signatures}}
+```
+
+After generating the signatures, we can actually send a transaction to spend the predicate funds. We use the `receiver` wallet as the recipient. In the transaction we have to provide both the predicate byte code and the required signatures. As we provide the correct data we receive the funds and verify that the amount is correct.
```
Comment (MujkicA) [inline docs/src/predicates/send-spend-predicate.md:41]: Typo + personal preference
```suggestion
After generating the signatures, we can send a transaction to spend the predicate funds. We use the `receiver` wallet as the recipient. We have to provide the predicate byte code and the required signatures. As we provide the correct data, we receive the funds and verify that the amount is correct.
```

**Review by MujkicA — COMMENTED** [review]
Just a couple of nits, maybe we could/should mention that signing over an array of zeroes is just for the sake of the example. 
Great example btw!

File: examples/predicates/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+    use fuels::prelude::Error;
+
+    #[tokio::test]
+    async fn predicate_example() -> Result<(), Error> {
+        use fuels::contract::predicate::Predicate;
+        use fuels::prelude::*;
+        use fuels::signers::fuel_crypto::SecretKey;
+
+        // ANCHOR: predicate_wallets
+        let secret_key1: SecretKey =
```
Comment (segfault-magnet) [inline examples/predicates/src/lib.rs:12]: A fun helper:
```rust
    fn wallets_from_keys<const T: usize>(keys: &[&str; T]) -> [LocalWallet; T] {
        keys.map(|hash| LocalWallet::new_from_private_key(hash.parse().unwrap(), None))
    }
    // ... later on
        let [wallet, wallet2, wallet3] = wallets_from_keys(&[
            "0x862512a2363db2b3a375c0d4bbbd27172180d89f23f2e259bac850ab02619301",
            "0x37fa81c84ccd547c30c176b118d5cb892bdb113e8e80141f266519422ef9eefd",
            "0x976e5c3fa620092c718d852ca703b6da9e3075b9f2ecb8ed42d9f746bf26aafb",
        ]);

```

Comment (segfault-magnet) [reply examples/predicates/src/lib.rs:12]: Albeit only now do I see the anchors. We probably keep things 'vanilla' here :) 

But I'll leave the comment be :) Might be useful in the future.

**Review by segfault-magnet — APPROVED** [review]

**Review by MujkicA — APPROVED** [review]
LGTM

**Review by Br1ght0ne — APPROVED** [review]
LGTM!

**Review by digorithm — APPROVED** [review]

**Pull request merged**